### PR TITLE
Adding FACES parser to datman

### PR DESCRIPTION
**Pull request by benselby**
Adding an old script for parsing OPT's FACES task to datman.

### CODE REVIEW

File: bin/dm_parse_faces.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python
+"""FACES Task Parser
+
+Converts eprime text file for FACES task into BIDS .tsv files.
+
+Usage: dm_parse_faces.py [options] <study> [<session>] [--output-dir=<out_dir>]
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:6]:
```suggestion
Usage:
    dm_parse_faces.py [options] <study>
    dm_parse_faces.py [options] <study> [<session>]
    dm_parse_faces.py [options] <study> [<session>] [--output-dir=<out_dir>]
```

File: bin/dm_parse_faces.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python
+"""FACES Task Parser
+
+Converts eprime text file for FACES task into BIDS .tsv files.
+
+Usage: dm_parse_faces.py [options] <study> [<session>] [--output-dir=<out_dir>] 
+
+Arguments:
+    <study>     Name of the study to process e.g. OPT
+    <session>   Datman name of session to process e.g. OPT01_UP1_UP10044_02_03
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:10]:
```suggestion
    <session>   Datman name of session to process e.g. OPT01_UP1_UP10004_02_01
```

Comment (kyjimmy) [reply bin/dm_parse_faces.py:10]: No data inside that participant's folder.

File: bin/dm_parse_faces.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+Options:
+    --debug         Set log level to debug
+    --output-dir    Specify an alternate output directory 
+    --dry-run       Don't actually write any output
+
+Outputs the trial number, onset time, accurracy, and reaction time values for each trial.
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:17]:
```suggestion
Outputs the trial number, onset time, accurracy, and reaction time values \
    for each trial.
```

Comment (kyjimmy) [reply bin/dm_parse_faces.py:17]: pep8 maximum characters issue

File: bin/dm_parse_faces.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+Usage: dm_parse_faces.py [options] <study> [<session>] [--output-dir=<out_dir>] 
+
+Arguments:
+    <study>     Name of the study to process e.g. OPT
+    <session>   Datman name of session to process e.g. OPT01_UP1_UP10044_02_03
+
+Options:
+    --debug         Set log level to debug
+    --output-dir    Specify an alternate output directory 
+    --dry-run       Don't actually write any output
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:15]:
```suggestion
    --output-dir=<out_dir>  Specify an alternate output directory
    --debug                 Set log level to debug
    --dry-run               Don't actually write any output
```

File: bin/dm_parse_faces.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+import re
+import os
+import glob
+
+import datman.config
+import datman.scanid
+
+logging.basicConfig(
+        level=logging.WARN, format="[%(name)s] %(levelname)s: %(message)s"
+        )
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:36]:
```suggestion
logging.basicConfig(level=logging.WARN,
                    format="[%(name)s] %(levelname)s: %(message)s")
```

File: bin/dm_parse_faces.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+logger = logging.getLogger(os.path.basename(__file__))
+
+_to_esc = re.compile(r'\s|()')
+def _esc_fname(match):
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:40]:
```suggestion


def _esc_fname(match):
```

File: bin/dm_parse_faces.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    return '\\' + match.group(0)
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:42]:
```suggestion


```

File: bin/dm_parse_faces.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        logger.setLevel(logging.DEBUG)
+
+    if dryrun:
+        logger.info("Dry run - will not write any output")
+
+    config = datman.config.config(study=study)
+
+    task_path = config.get_path("task")
+    nii_path = config.get_path("nii")
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:139]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    for ses in sessions:
+        logger.info(f"Parsing {ses}...")
+
+        try:
+            ident = datman.scanid.parse(ses)
+        except datman.scanid.ParseException:
+            logger.error(
+                f"Skipping task folder with malformed ID {ses}"
+            )
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:155]:
```suggestion
            logger.error(f"Skipping task folder with malformed ID {ses}")
```

File: bin/dm_parse_faces.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            continue
+        ses_path = os.path.join(task_path, ses)
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:157]:
```suggestion
        ses_path = os.path.join(task_path, ses)
```

File: bin/dm_parse_faces.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        for eprimefile in glob.glob(ses_path + '/*.txt'):
+            logger.info(f"Found file: {eprimefile}")
+            try:
+                eprime = read_eprime(eprimefile)
+            except UnicodeError as e:
+                logger.error(f"Cannot parse {eprimefile}: {e}")
+                continue
+            # tag the trials to obtain the data for each trial
+            taglist = find_all_data(eprime, "Procedure: TrialsPROC\r\n")
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:167]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+            # tag the trials to obtain the data for each trial
+            taglist = find_all_data(eprime, "Procedure: TrialsPROC\r\n")
+            
+            if not taglist:
+                logger.error(f"No trials found for {ses} - skipping")
+                continue
+
+            trial_start = np.empty([len(taglist)], dtype=int)
+            trial_end = np.empty([len(taglist)], dtype=int)
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:174]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+            trial_start = np.empty([len(taglist)], dtype=int)
+            trial_end = np.empty([len(taglist)], dtype=int)
+            
+            for i, ind_trial_proc in enumerate(taglist):
+                if (i < (len(taglist)) - 1):
+                    trial_end[i] = taglist[i + 1][0]
+                elif (i == (len(taglist)) - 1):
+                    trial_end[i] = len(eprime) - 1
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:180]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+            trial_end = np.empty([len(taglist)], dtype=int)
+            
+            for i, ind_trial_proc in enumerate(taglist):
+                if (i < (len(taglist)) - 1):
+                    trial_end[i] = taglist[i + 1][0]
+                elif (i == (len(taglist)) - 1):
+                    trial_end[i] = len(eprime) - 1
+        
+                trial_start[i] = ind_trial_proc[0]
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:182]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+            for i, ind_trial_proc in enumerate(taglist):
+                if (i < (len(taglist)) - 1):
+                    trial_end[i] = taglist[i + 1][0]
+                elif (i == (len(taglist)) - 1):
+                    trial_end[i] = len(eprime) - 1
+        
+                trial_start[i] = ind_trial_proc[0]
+        
+            trial_blocks = [eprime[s:e] for s, e in zip(trial_start, trial_end)]
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:184]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+        
+            trial_blocks = [eprime[s:e] for s, e in zip(trial_start, trial_end)]
+        
+            entries = []
+            for b in trial_blocks:
+                entries.append({
+                    'onset':
+                    get_event_value(b, 'StimSlide.OnsetTime:'),
+                    'duration':
+                    get_event_value(b, 'StimSlideOnsetToOnsetTime:'),
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:191]:
```suggestion
                    get_event_value(b, 'StimSlide.OnsetToOnsetTime:'),
```

File: bin/dm_parse_faces.py [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+                    'response_time':
+                    get_event_value(b, 'StimSlide.RT:'),
+                    'accuracy':
+                    get_event_value(b, 'StimSlide.ACC:'),
+                    'correct_response':
+                    map_response(get_event_value(b, 'CorrectResponse:')),
+                    'participant_response':
+                    map_response(get_event_value(b, 'StimSlide.RESP:'))
+                })
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:203]:
```suggestion

```

File: bin/dm_parse_faces.py [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+                    .astype({
+                        "correct_response": "Int64",
+                        "participant_response": "Int64",
+                        "accuracy": "Int64"
+                    })
+        
+            log_head, log_tail = os.path.split(eprimefile)
+        
+            sub_id = os.path.basename(log_head)
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:221]:
```suggestion
                "onset": np.float,
                "duration": np.float,
                "response_time": np.float,
                "accuracy": np.float,
                "correct_response": np.float,
                "participant_response": np.float
            })\
                .astype({
                    "correct_response": "Int64",
                    "participant_response": "Int64",
                    "accuracy": "Int64"
                })

            # log_head, log_tail = os.path.split(eprimefile)
            # sub_id = os.path.basename(log_head)

```

File: bin/dm_parse_faces.py [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+            
+            if not out_dir:
+                out_dir = os.path.join(nii_path, 
+                            ident.get_full_subjectid_with_timepoint() 
+                          )
+        
+            file_name = os.path.join(out_dir, f"{ses}_FACES.tsv") 
+            if not os.path.exists(os.path.dirname(file_name)):
+                os.makedirs(os.path.dirname(file_name))
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:230]:
```suggestion
                out_dir = os.path.join(nii_path,
                                       ident.get_full_subjectid_with_timepoint()
                                       )

            file_name = os.path.join(out_dir, f"{ses}_FACES.tsv")

```

File: bin/dm_parse_faces.py [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+            file_name = os.path.join(out_dir, f"{ses}_FACES.tsv") 
+            if not os.path.exists(os.path.dirname(file_name)):
+                os.makedirs(os.path.dirname(file_name))
+            
+            if not dryrun:
+                logger.info(f"Saving output to {file_name}")
+                data.to_csv(file_name, sep='\t', index=False)
+            else:
+                logger.info(f"Dry run - would save to {file_name}")
+
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:236]:
```suggestion


```

File: bin/dm_parse_faces.py [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+            if not out_dir:
+                out_dir = os.path.join(nii_path, 
+                            ident.get_full_subjectid_with_timepoint() 
+                          )
+        
+            file_name = os.path.join(out_dir, f"{ses}_FACES.tsv") 
+            if not os.path.exists(os.path.dirname(file_name)):
+                os.makedirs(os.path.dirname(file_name))
+            
+            if not dryrun:
```
Comment (kyjimmy) [inline bin/dm_parse_faces.py:231]:
```suggestion
            if not dryrun:
                os.makedirs(os.path.dirname(file_name), exist_ok=True)
```

**Review by kyjimmy — CHANGES_REQUESTED** [review]
Mostly style changes with a few minor changes with `docopt` usage and the extraction of `duration`.

File: bin/dm_parse_faces.py (hunk 2 above)
Comment (benselby) [reply bin/dm_parse_faces.py:10]: There is task data in OPT/task in the archive for this pt.

**Comment by jerdra** [comment]
Will close this PR and take over, thanks for the comments @kyjimmy, am integrating them rn and will re-open with your suggestions!

**Pull request closed without merging**